---
title: "Deduplicate gene expression matrix on per-sample basis"
author: Carolin Turner
---

##################################################
README:
There are replicate gene names due to alternative sequences and error in the reference assembly. Here, we deduplicate an annotated gene expression matrix, retaining for each sample the highest expression value among replicate genes.

The input matrix should be a subset of the TPM or CPM matrix that is the default output of our RNAseq processing pipeline step3 script, filtered for genes of interest (e.g. all protein-coding genes, or all constituent genes of an integrated module list).

The output matrix will only contain GeneSymbol annotation. Ensembl Gene ID is dropped because a different ID may be retained for different samples. Gene biotype is dropped because very occasionally two Ensembl IDs with the same gene name annoation will have a different gene biotype.

In addition to the output matrix, this script will create two folders in your working directory, called 'Files_per_sample' and 'Files_per_sample_dedup', which can be deleted after the entire script has been run.

###################################################

Load required packages
```{r}
rm(list=ls())
library(dplyr)
library(textshape)
```

Change here the name of your working directory (where input file is located and output file will be saved), your input filename, and desired output matrix name. 
```{r}
working.directory <- "/home/afonso/OneDriveUCL/UCL-Work2/PGU/PGU2_RNAseq_Weight/PGU2_CWeight"
input.file <- "Annotatedexample_tpm_PC0.001_log2.csv"
output.file <- "example_tpm_PC0.001_log2_genesymbol_dedup.csv"
```

Create output directories
```{r}
wd <- working.directory
dir1 <- paste0(wd,"/Files_per_sample")
dir.create(dir1)

dir2 <- paste0(wd, "/Files_per_sample_dedup")
dir.create(dir2)
```

Read in input file, save list of gene symbols, remove annotations from expression matrix, and save list of samples
```{r}
setwd(wd)

tpm <- read.csv(input.file)
genes <- as.data.frame(tpm %>% dplyr::select(external_gene_name))
tpm <- tpm %>% dplyr::select(-c(ensembl_gene_id,external_gene_name,description,gene_biotype))
names <- colnames(tpm) 
```

Loop through samples, merge with list of genes and save single sample files in first output directory
```{r}
setwd(dir1)
  
for (i in 1:length(names)){
  sample <- tpm[,i]
  merged <- cbind(genes,sample)
  write.csv(merged,file=paste0(names[i],".csv"),row.names = F)
}
```

Loop through single sample files, deduplicate based on highest expression and save deduplicated single sample files in second output directory
```{r}
setwd(dir1)

allfiles <- list.files(pattern = ".csv")

for (i in 1:length(allfiles)){
  dat <- read.csv(allfiles[i],header=TRUE)
  dat <- dat[order(dat$external_gene_name, -(dat$sample)), ]
  dat <- dat[!duplicated(dat$external_gene_name), ]
 
  nam <- gsub('\\.csv','',allfiles[i])
  colnames(dat)[2] <- nam
  
  write.csv(dat,file=paste0(dir2,"/",nam,"_dedup.csv"),row.names = F)
}
```

Merge deduplicated single sample files into one matrix and save in working directory
```{r}
# read in first deduplicated sample file and set first column name to GeneSymbol
setwd(dir2)
allfiles.dedup <- list.files(pattern = ".csv")

df <- read.csv(allfiles.dedup[1], header=T) 
colnames(df)[1] <- "GeneSymbol" 

# loop through all remaining files and merge with the first one, using GeneSymbol as common column
for(f in 2:length(allfiles.dedup)){ 
  dat <- read.csv(allfiles.dedup[f],header=TRUE)
  colnames(dat)[1] <- "GeneSymbol"
  df <- merge(df,dat,by="GeneSymbol")
}

# order columns alphabetically
df <- df[,order(colnames(df))]

# use GeneSymbol as rownames
df <- df %>% column_to_rownames("GeneSymbol")

# write to file
setwd(wd)
write.csv(df,output.file)
```

